core/nightshadev2/src/nightshade.rs

```diff
@@ -0,0 +1,265 @@
+    fn cmp(&self, other: &Self) -> Ordering {
+        if self.confidence0 != other.confidence0 {
+            return if self.confidence0 > other.confidence0 {
+                Ordering::Greater
+            } else {
+                Ordering::Less
+            };
+        }
+
+        if self.endorses != other.endorses {
+            return if self.endorses < other.endorses {
+                Ordering::Greater
+            } else {
+                Ordering::Less
+            };
+        }
+
+        if self.confidence1 != other.confidence1 {
+            return if self.confidence1 > other.confidence1 {
+                Ordering::Greater
+            } else {
+                Ordering::Less
+            };
+        }
+
+        Ordering::Equal
+    }
+}
+
+impl BareState {
+    fn new(endorses: AuthorityId) -> Self {
+        Self {
+            endorses,
+            confidence0: 0,
+            confidence1: 0,
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct SignedState {
+    signature: BLSSignature,
+    parent: Vec<BareState>,
+}
+
+impl SignedState {
+    fn new() -> Self {
+        Self {
+            signature: 0,
+            parent: vec![],
+        }
+    }
+
+    fn update(&mut self, state: &State) {
+        // TODO: Update self.signature using state.get_signature
+        self.parent.push(state.bare_state.clone());
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct State {
+    bare_state: BareState,
+
+    // TODO: Proof might be empty at the beginning of consensus. Use enum instead?
+    proof0: Option<SignedState>,
+    proof1: Option<SignedState>,
+}
+
+impl State {
+    fn new(endorses: AuthorityId) -> Self {
+        Self {
+            bare_state: BareState::new(endorses),
+            proof0: None,
+            proof1: None,
+        }
+    }
+
+    /// Create new State with increased confidence using some proof
+    fn increase_confidence(&self, proof: SignedState) -> Self {
+        Self {
+            bare_state: BareState {
+                endorses: self.bare_state.endorses,
+                confidence0: self.bare_state.confidence0 + 1,
+                confidence1: self.bare_state.confidence1,
+            },
+            proof0: Some(proof),
+            proof1: self.proof1.clone(),
+        }
+    }
+
+    fn can_commit(&self) -> bool {
+        self.bare_state.confidence0 >= self.bare_state.confidence1 + COMMIT_THRESHOLD
+    }
+
+    fn verify(&self) -> bool {
+        true
+    }
+
+    fn get_signature(&self) -> BLSSignature {
+        0
+    }
+
+    fn endorses(&self) -> usize {
+        self.bare_state.endorses
+    }
+}
+
+impl PartialOrd for State {
+    fn partial_cmp(&self, other: &State) -> Option<Ordering> {
+        self.bare_state.partial_cmp(&other.bare_state)
+    }
+}
+
+impl Ord for State {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.bare_state.cmp(&other.bare_state)
+    }
+}
+
+
+fn update_confidence1(state0: &mut State, state1: &State){
```

Comment:
should we update confidence1 if state0 = (A, 5, 3) and state1 = (A, 4, 4) ?

Reply:
Fixed [here](https://github.com/nearprotocol/nearcore/pull/549/commits/20ee4477929432440e24d7eabb4f45b6622e8ff5#diff-eb6ef29898d59b2505e1e8a1c6dfab43R158)